### ISSUE TITLE
Multiple Maps - One loaded on page load, the rest via AJAX

### ISSUE THREAD

**Issue by mckelveydigital** [bradcornford/Googlmapper#102]
I am creating an application that shows a map on page load, which I have no issues with, however when I attempt to use AJAX to load in more content (including a map), I get a console error explaining that the Google Maps API has been included more than once. 

In my view that is being returned by AJAX I am using {!! Mapper::render() !!}, and my controller for the AJAX request has Mapper::map($lat, $lng, ['zoom' => 15, 'center' => true, 'marker' => true]);

I think the issue is Mapper::render() outputs the Google Maps include each time it is used. Is there any way to just render the map to the page, without the Google Maps API script?

Many thanks in advance!

**Comment by mckelveydigital** [comment]
Kept going at it myself and ended up getting it working. The issue wasn't only due to the double include of the Google Maps API, it was also due to the individual map script not being executed after the AJAX request and because the new map had the same ID as the existing one.

To fix the double include of the Google Maps API, I removed the following line from the mapper.blade.php view -

`@include('googlmapper::javascript')`

I then included my own link to google maps as well as the markerclusterer.js file in my layout view.

To fix the issue with the individual map script not being executed I had to first give each map canvas its own ID, which in turn ment I had to update the `initialize_{!! $id !!}()` function as well to reference the new ID. I used `<?php $new_id = rand(5, 10); ?>` to generate a random number between 1 and 10 and changed the ID of the map canvas and the name of the initialize function to reflect this.

It's worth noting that even with all this, it still wouldn't work until I added brackets to the end of the function name in the last line of the script as well, these were missing?

`google.maps.event.addDomListener(window, 'load', initialize_{!! $new_id !!}());`
instead of
`google.maps.event.addDomListener(window, 'load', initialize_{!! $new_id !!});`

My full modified mapper.blade.php file is shown below

`<?php $new_id = rand(5, 10); ?>`

`<div id="map-canvas-{!! $new_id !!}" class="map-canvas" style="height: 100%; margin: 0; padding: 0;"></div>`
`<script id="script_{!! $new_id !!}" type="text/javascript">`

	`var maps = [];

	function initialize_{!! $new_id !!}() {
		var bounds = new google.maps.LatLngBounds();
		var infowindow = new google.maps.InfoWindow();
		var position = new google.maps.LatLng({!! $options['latitude'] !!}, {!! $options['longitude'] !!});

		var mapOptions_{!! $id !!} = {
			@if ($options['center'])
				center: position,
			@endif
			mapTypeId: google.maps.MapTypeId.{!! $options['type'] !!},
			disableDefaultUI: @if (!$options['ui']) true @else false @endif,
			scrollwheel: @if ($options['scrollWheelZoom']) true @else false @endif,
			fullscreenControl: @if ($options['fullscreenControl']) true @else false @endif,
		};

		var map_{!! $id !!} = new google.maps.Map(document.getElementById('map-canvas-{!! $new_id !!}'), mapOptions_{!! $id !!});
		map_{!! $id !!}.setTilt({!! $options['tilt'] !!});

		var markers = [];
		var infowindows = [];
		var shapes = [];

		@foreach ($options['markers'] as $key => $marker)
			{!! $marker->render($key, $view) !!}
		@endforeach

		@foreach ($options['shapes'] as $key => $shape)
			{!! $shape->render($key, $view) !!}
		@endforeach

		@if ($options['overlay'] == 'BIKE')
			var bikeLayer = new google.maps.BicyclingLayer();
			bikeLayer.setMap(map_{!! $id !!});
		@endif

		@if ($options['overlay'] == 'TRANSIT')
			var transitLayer = new google.maps.TransitLayer();
			transitLayer.setMap(map_{!! $id !!});
		@endif

		@if ($options['overlay'] == 'TRAFFIC')
			var trafficLayer = new google.maps.TrafficLayer();
			trafficLayer.setMap(map_{!! $id !!});
		@endif

		var idleListener = google.maps.event.addListenerOnce(map_{!! $id !!}, "idle", function () {
			map_{!! $id !!}.setZoom({!! $options['zoom'] !!});

			@if (!$options['center'])
				map_{!! $id !!}.fitBounds(bounds);
			@endif

			@if ($options['locate'])
				if (typeof navigator !== 'undefined' && navigator.geolocation) {
					navigator.geolocation.getCurrentPosition(function (position) {
						map_{!! $id !!}.setCenter(new google.maps.LatLng(position.coords.latitude, position.coords.longitude));
					});
				}
			@endif
		});

		@if (isset($options['eventBeforeLoad']))
			{!! $options['eventBeforeLoad'] !!}
		@endif

		@if (isset($options['eventAfterLoad']))
			google.maps.event.addListenerOnce(map_{!! $id !!}, "tilesloaded", function() {
				{!! $options['eventAfterLoad'] !!}
			});
		@endif

		@if ($options['cluster'])
			var markerClusterOptions = {
				imagePath: '{!! $options['clusters']['icon'] !!}',
				gridSize: {!! $options['clusters']['grid'] !!},
				maxZoom: @if ($options['clusters']['zoom'] === null) null @else {!! $options['clusters']['zoom'] !!} @endif,
				averageCenter: @if ($options['clusters']['center'] === true) true @else false @endif,
				minimumClusterSize: {!! $options['clusters']['size'] !!}
			};
			var markerCluster = new MarkerClusterer(map_{!! $id !!}, markers, markerClusterOptions);
		@endif

		maps.push({
			key: {!! $id !!},
			markers: markers,
			infowindows: infowindows,
			map: map_{!! $id !!},
			shapes: shapes
		});
		
		console.log("here");
		
	}

	google.maps.event.addDomListener(window, 'load', initialize_{!! $new_id !!}());`
	
`</script>`

I don't know if this is the best solution to this, but it worked for me. If there is a better or easier way please let me know.

**Comment by bradcornford** [comment]
Hi there,

It seems like you have found a solution yourself, but there may be an easier method to achieve this. I'll investigate this. 

One thing that may simply your solution is to keep the javascript include, however in your controller for the AJAX maps, you can simple add `View::share('javascript', true);` to disable the rendering of the subsequent javascript includes.

**Comment by mckelveydigital** [comment]
Hi,

I thought there might be an easier way to stop the javascript includes, I will implement that solution instead. 

Thanks for your help, great work!

**Comment by bradcornford** [comment]
The only other problem i can foresee with your solution is that the more times you load a map via AJAX, you may encounter a problem with the random number generation picking the same number again.

You could also just overwrite the `id` variable itself, which would make it a little simpler and not diverge from master so much.

As for needing to add brackets to the initialize call, according to Google's documentation: https://developers.google.com/maps/documentation/javascript/events#DomEvents it accepts a string as a function name, or a function itself, so both should work equally.

I'll also look at implementing a method in the future `renderWithoutJavascript`, which will make the process you are doing easier. 

I'll also look at adding the possibility of adding a method parameter of prefix, to the render methods to allow the above type behaviour you are doing natively. That way you could pass a prefix of `time()` for example, which would ensure that all your maps are uniquely indexed for the current request.

**Comment by mckelveydigital** [comment]
Yeah I can see what you mean, I suppose widening the parameters of the rand function to something a bit more drastic could solve that.

That is strange about the brackets not being required, of course the code worked for the initial map without the brackets so this is true. The only thing I can think of there is when the script tag is rendered to the page with AJAX, it mustn't pick up on the fact that it is a function and not a string? Dropping the brackets now still causes the error on my end? Could be an issue with the Google Maps API itself.

I researched loading script elements with AJAX and realised that its more complicated than I thought it would be. A lot of people are suggesting running an eval() function on the AJAX data in order to run any functions in javascript elements loaded via AJAX, as simply adding them to the page doesn't trigger them for some reason.

The features you suggested sound great, what I am doing probably isn't that common but I am sure someone else will benefit from that further down the line.

**Comment by bradcornford** [comment]
Rather than generating a larger random number, perhaps something like: 

```
<?php $new_id = time() . '_' . $id; ?>
```

**Comment by mckelveydigital** [comment]
Much better solution thanks!

Yeah I didn't have to use it as adding the brackets made it work anyway. But yes eval is a nightmare!

**Comment by reymark-torres** [comment]
Is there an available solution to this?
How to properly re render the Google Map via AJAX?
I have a search function to re-render the Google Map to display new markers based in the input of the user.

I will really appreciate your response.
Thank you very much

**Comment by githubgobi** [comment]
Multiple marker options available in single map? @reymark-torres

**Comment by mrkalmdn** [comment]
@reymark-torres, I've used before the code of @mckelveydigital. It will render the map via AJAX request.
